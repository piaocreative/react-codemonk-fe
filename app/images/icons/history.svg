<svg width="18" height="18" viewBox="0 0 18 18" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M1 9C1 13.4186 4.58141 17 9 17C13.4186 17 17 13.4186 17 9C17 4.58141 13.4186 1 9 1C6.25007 1 3.8244 2.38717 2.3845 4.5" stroke="#04004E" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M11.9679 11.5448L8.70732 9.5997V5.4077" stroke="#04004E" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
<path d="M2.10001 2L2.10001 4.82843H4.92843" stroke="#04004E" stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round"/>
</svg>
