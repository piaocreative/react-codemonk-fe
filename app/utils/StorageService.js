import base from 'base-64';

/**
 * This class used for Store value on localStorage of browser.
 * @author Innovify
 */

const StorageService = {
  /**
   * check value exists on localStorage of browser.
   * @author Innovify
   * @Developer Ankit Gujarati
   */
  exists(key) {
    return this.get(key);
  },
  /**
   * set value from localStorage of browser.
   * @author Innovify
   * @Developer Ankit Gujarati
   */
  set(key, value, opts = {}) {
    let storedValue = value;
    if (opts.stringify) storedValue = JSON.stringify(storedValue);
    if (opts.hash) storedValue = base.encode(storedValue);
    try {
      if (opts.keepMe) {
        const d = new Date();
        d.setTime(d.getTime() + 30 * 24 * 60 * 60 * 1000);
        const expires = `expires=${d.toUTCString()}`;
        document.cookie = `${key}=${JSON.stringify({
          opts,
          body: storedValue,
        })};${expires};path=/`;
      } else {
        document.cookie = `${key}=${JSON.stringify({
          opts,
          body: storedValue,
        })};path=/`;
      }
    } catch (err) {
      throw err;
    }
  },
  /**
   * get value from localStorage of browser.
   * @author Innovify
   */
  get(key) {
    // if (!this.exists(key)) return false;
    const name = `${key}=`;
    const decodedCookie = decodeURIComponent(document.cookie);
    const ca = decodedCookie.split(';');
    // eslint-disable-next-line no-plusplus
    for (let i = 0; i < ca.length; i++) {
      let c = ca[i];
      while (c.charAt(0) === ' ') {
        c = c.substring(1);
      }
      if (c.indexOf(name) === 0) {
        const data = JSON.parse(c.substring(name.length, c.length));
        let { body } = data;
        if (data.opts && data.opts.hash) body = base.decode(body);
        if (data.opts && data.opts.stringify) body = JSON.parse(body);
        return body;
      }
    }
    return false;
    /* const data = JSON.parse(localStorage.getItem(key));
    let { body } = data;
    if (data.opts && data.opts.hash) body = base.decode(body);
    if (data.opts && data.opts.stringify) body = JSON.parse(body);
    return body; */
  },
  /**
   * delete value from localStorage of browser.
   * @author Innovify
   */
  delete(key, opts = {}) {
    const d = new Date();
    d.setTime(d.getTime() + 0);
    const expires = `expires=${d.toUTCString()}`;
    document.cookie = `${key}=${JSON.stringify({
      opts,
      body: '',
    })};${expires};path=/`;
    // localStorage.removeItem(key);
  },
  /**
   * clear localStorage of browser.
   * @author Innovify
   */
  clear() {
    this.delete('signupStep');
    this.delete('apiSignupStep');
    this.delete('userVersion');
    this.delete('user');
    this.delete('recentSearch');
    this.delete('jwtToken');
    this.delete('location');
    this.delete('unAuthorize');
    this.delete('userPhoneCountry');
    this.delete('registerType');
    this.delete('userEmail');
    this.delete('userType');
    this.delete('agencyLogo');
    this.delete('pageNumber');
    this.delete('userPhoneNumber');
    this.delete('profilePicture');
    this.delete('firstName');
    this.delete('userPhone');
    this.delete('Notification');
    this.delete('NotificationType');
    this.delete('talentProjectPageNumber');
    this.delete('talentProjectFilter');
    this.delete('clientProjectPageNumber');
    this.delete('clientProjectStatus');
    this.delete('isPaymentSkipped');
    this.delete('agencyMyTeamPageNumber');
    this.delete('adminAgencyPageNumber');
    this.delete('adminAgencyFilterObject');
    this.delete('adminBriefDetailID');
    this.delete('clientsFilterObject');
    this.delete('clientsPageNumber');
    this.delete('interviewFilterObject');
    this.delete('interviewPageNumber');
    this.delete('adminProjectDetailsBriefPageNumber');
    this.delete('projectFilterObject');
    this.delete('projectPageNumber');
    this.delete('talentFilterObject');
    this.delete('talentPageNumber');
    this.delete('agencyQuotePage');
    this.delete('selectedRoleType');
    this.delete('filterObject');
    this.delete('proxyType');
    this.delete('proxyToken');
    this.delete('talentBriefPage');
    this.delete('talentBriefFilter');
    this.delete('clientBriefDetailsPageNumber');
    this.delete('clientBriefPage');
    this.delete('clientBriefFilter');
    this.delete('clientID');
    this.delete('adminTalentSearch');
    this.delete('adminAgencySearch');
    this.delete('adminClientSearch');
    this.delete('isLoginViaAdmin');
    this.delete('userBillingDetailsObj');
  },
};

export default StorageService;
