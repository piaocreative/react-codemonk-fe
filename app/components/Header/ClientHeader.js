import React from 'react';
import { Link } from 'react-router-dom';
import { userExists } from 'utils/Helper';
import { FormattedMessage } from 'react-intl';
import get from 'lodash/get';
import { toast } from 'react-toastify';
import { Dropdown, DropdownItem } from 'reactstrap';
import { defaultProps, propTypes } from 'containers/proptypes';
import LogoutLink from 'components/LogoutLink';
import { A, H1 } from 'components';
import SVG from 'react-inlinesvg';
import StorageService from 'utils/StorageService';
import Emitter from 'utils/emitter';
import history from 'utils/history';
import {
  userIcon,
  settingIcon,
  logoutIcon,
  backArrowIcon,
  clientsGIcon,
  teamGIcon,
  quotesGIcon,
  TYPEFORM_URL,
  routes,
} from 'containers/App/constants';
import containerMessage from 'containers/messages';
import { redirectTo } from 'containers/App/utils';
import ClientLogo from 'components/Brand/ClientLogo';
import AddProject from 'containers/Admin/Projects/AddProject';
import AddBrief from 'containers/Client/AddBrief';
import { handleBackToAdmin } from 'containers/Admin/ProxyLogin/utils';
import { getSkills } from 'containers/Auth/utils';
import Notifications from 'containers/Notifications';
import Autocomplete from 'containers/TalentListingPage/AutoComplete';
import ToastifyMessage from 'components/ToastifyMessage';
import {
  UserImg,
  HeaderNav,
  FixedNavbar,
  NavbarContainer,
  UserProfileToggle,
  UserProfileMenu,
  HeaderActions,
  HamburgerToggle,
  SecondaryNav,
} from './header-style';
import messages from './messages';

const getUsrDetail = () => ({
  isSignupStep: StorageService.get('signupStep') || '',
});

export class ClientHeader extends React.Component {
  constructor(props) {
    super(props);
    const profilePicture = StorageService.get('profilePicture');
    this.state = {
      profilePicture,
      skillsList: [],
      skillsArray: [],
    };
  }

  setConstructor = () => {
    const userType = StorageService.get('userType');
    const proxyType = StorageService.get('proxyType');
    const profilePicture = StorageService.get('profilePicture');
    const stateData = {
      dropdownOpen: false,
      hireTalentDropdownOpen: false,
      isSignIn: userExists(),
      isOpen: false,
      ...getUsrDetail(),
      userType,
      proxyType,
      badgeConfiguration: {},
      profilePicture,
    };
    this.setState({ ...stateData });
  };

  componentDidMount() {
    this.setConstructor();
    getSkills(this.setSkills);

    Emitter.on('proxyLoginClient', proxyLoginClient => {
      if (proxyLoginClient) {
        this.setConstructor();
      }
    });
    Emitter.on('profilePicture', profilePicture => {
      let url = '';
      if (profilePicture) {
        url = `${profilePicture}?_t=${new Date().getTime()}`;
      }
      this.setState({
        profilePicture: url,
      });
      StorageService.set('profilePicture', url, { hash: true });
    });

    Emitter.on('proxyBackToAdmin', proxyBackToAdmin => {
      if (proxyBackToAdmin) {
        this.setState({ dropdownOpen: false }, () => {
          this.setConstructor();
        });
      }
    });
    Emitter.on('badgeConfigurationUpdated', badgeConfigurationUpdated => {
      this.setState({ badgeConfiguration: badgeConfigurationUpdated });
    });
  }

  setSkills = response => {
    if (get(response, 'status')) {
      this.setState({ skillsList: response.data });
    } else {
      toast.error(<ToastifyMessage message={get(response, 'message')} type="error" />, { className: 'Toast-error' });
    }
  };

  componentWillUnmount() {
    Emitter.off('proxyLoginClient');
    Emitter.off('proxyBackToAdmin');
    Emitter.off('profilePicture');
    Emitter.off('badgeConfigurationUpdated');
  }

  /**
   * call toggle on tab change
   * @author Innovify
   */
  dropdownToggle = () => {
    this.setState(prevState => ({
      dropdownOpen: !prevState.dropdownOpen,
    }));
  };

  menuToggle = () => {
    this.setState(prevState => ({
      isOpen: !prevState.isOpen,
    }));
  };

  searchSkillsValueChangedTemp = (data, clearFilter) => {
    const { value } = data;
    const filterObjectGet = StorageService.get('filterObject');
    let filterObject;
    try {
      filterObject = JSON.parse(filterObjectGet) || { skillsArray: [] };
    } catch {
      filterObject = { skillsArray: [] };
    }
    const skillsArray = filterObject.skillsArray || [];
    const newSkillsArray = clearFilter ? skillsArray.filter(i => i.value !== value) : [value].concat(skillsArray);
    this.storeSelectedFilterValue('skillsArray', newSkillsArray.map(val => val));
    this.setReservedFilter(newSkillsArray.map(val => val));
    if (!(history && history.location && history.location.pathname === '/client/talent-listing')) {
      redirectTo(history, '/client/talent-listing');
    }
  };

  setReservedFilter = skillsArray => {
    this.setState({ skillsArray });
  };

  storeSelectedFilterValue = (keyName, KeyValue) => {
    const filterObjectGet = StorageService.get('filterObject');
    const filterObject = JSON.parse(filterObjectGet) || {};
    filterObject[keyName] = KeyValue;
    StorageService.set('filterObject', JSON.stringify(filterObject));
    Emitter.emit('filterObject');
  };

  getFormattedPath = pathName => {
    let output = '';
    output = pathName.split('/', 3).join('/');
    return output;
  };

  getPageTitle = () => {
    let output = '';
    const pathName = get(history.location, 'pathname');
    const formattedPath = this.getFormattedPath(pathName);
    output = routes.find(obj => obj.paths.find(i => this.getFormattedPath(i.path) === formattedPath)).headerTitle;
    return output;
  };

  render() {
    const {
      profilePicture,
      userType,
      isSignIn,
      dropdownOpen,
      isSignupStep,
      isOpen,
      proxyType,
      badgeConfiguration,
      skillsList,
      skillsArray,
    } = this.state;
    return (
      <HeaderNav className={isOpen && 'sidebarOpen'}>
        <FixedNavbar className="client-header">
          <NavbarContainer fluid>
            <div className="d-flex align-items-center navbar-expand-md">
              <ClientLogo />
              <H1 className="mb-0">{this.getPageTitle()}</H1>
              {isSignIn && isSignupStep >= '3' && <HamburgerToggle onClick={this.menuToggle} />}
            </div>

            {isSignIn && userType === '2' && (
              <div className="d-flex">
                {isSignupStep >= '3' && (
                  <>
                    <div>
                      <Autocomplete
                        customClass="skill-search order-lg-0 order-1"
                        options={skillsList.filter(v => skillsArray.filter(i => i.value === v).length === 0)}
                        label=""
                        filterType="skillSearch"
                        placeholder="Search by skills"
                        valueChanged={this.searchSkillsValueChangedTemp}
                      />
                    </div>
                    <HeaderActions className="d-flex">
                      <AddBrief
                        type="add"
                        btnTitle={containerMessage.createBrief.defaultMessage}
                        btnClassName="top-0 btn btn-sm btn-outline btn-plus"
                        withAddIcon
                        isClientHeader
                      />
                      <AddProject
                        {...this.props}
                        btnTitle={containerMessage.btnNewProject.defaultMessage}
                        btnClassName="ms-3 btn btn-sm btn-plus btn-primary"
                        withAddIcon
                        isClientProject
                        isClientHeader
                      />
                    </HeaderActions>
                    <SecondaryNav>
                      <Notifications newNotification={get(badgeConfiguration, 'newNotification')} />
                    </SecondaryNav>
                  </>
                )}
                <Dropdown isOpen={dropdownOpen} toggle={this.dropdownToggle}>
                  <UserProfileToggle>
                    {profilePicture ? (
                      <UserImg src={profilePicture} className="rounded-circle" alt="user-profile" />
                    ) : (
                      <SVG src={userIcon} />
                    )}
                  </UserProfileToggle>
                  <UserProfileMenu right>
                    <div className="inner-block">
                      {isSignupStep >= '3' && (
                        <>
                          <DropdownItem tag={Link} to="/client/about-you">
                            <SVG src={userIcon} />
                            <FormattedMessage {...messages.menuItemMyProfile} />
                          </DropdownItem>
                          <DropdownItem tag={Link} to="/client/account-settings">
                            <SVG src={settingIcon} />
                            <FormattedMessage {...messages.menuItemAccountSetting} />
                          </DropdownItem>
                          <DropdownItem className="d-md-none" tag={Link} to="/client/job-briefs">
                            <SVG className="client-icon" src={clientsGIcon} />
                            <FormattedMessage {...messages.menuItemPostBrief} />
                          </DropdownItem>
                          <DropdownItem className="d-md-none" tag={Link} to="/client/talent-listing">
                            <SVG className="client-icon" src={clientsGIcon} />
                            <FormattedMessage {...messages.menuItemSearchTalent} />
                          </DropdownItem>
                          <DropdownItem className="d-md-none text-decoration-none" tag={A} target="_blank" href={TYPEFORM_URL}>
                            <SVG src={teamGIcon} />
                            <FormattedMessage {...messages.buttonBuildTeam} />
                          </DropdownItem>
                          <DropdownItem className="d-md-none" tag={Link} to="/client/start-project">
                            <SVG src={quotesGIcon} />
                            <FormattedMessage {...messages.menuRequestQuote} />
                          </DropdownItem>
                        </>
                      )}
                      <LogoutLink>
                        <SVG src={logoutIcon} />
                        <FormattedMessage {...messages.menuItemLogout} />
                      </LogoutLink>
                      {/* proxyLogin Return to admin */}
                      {proxyType && (
                        <DropdownItem onClick={handleBackToAdmin}>
                          <SVG src={backArrowIcon} />
                          <FormattedMessage {...messages.menuItemBackToAdmin} />
                        </DropdownItem>
                      )}
                    </div>
                  </UserProfileMenu>
                </Dropdown>
              </div>
            )}

            {/* admin profile dropdown */}
            {isSignIn && userType === '4' && (
              <div className="d-flex">
                <Dropdown isOpen={dropdownOpen} toggle={this.dropdownToggle}>
                  <UserProfileToggle>
                    <SVG src={userIcon} />
                  </UserProfileToggle>
                  <UserProfileMenu right>
                    <div className="inner-block">
                      <DropdownItem tag={LogoutLink}>
                        <SVG src={logoutIcon} />
                        <FormattedMessage {...messages.menuItemLogout} />
                      </DropdownItem>
                    </div>
                  </UserProfileMenu>
                </Dropdown>
              </div>
            )}
          </NavbarContainer>
        </FixedNavbar>
      </HeaderNav>
    );
  }
}

ClientHeader.defaultProps = defaultProps;
ClientHeader.propTypes = propTypes;

export default ClientHeader;
