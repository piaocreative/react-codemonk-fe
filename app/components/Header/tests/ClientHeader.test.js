import React from 'react';
import { shallow } from 'enzyme';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import { IntlProvider } from 'react-intl';
import { ConnectedRouter } from 'connected-react-router/immutable';
import { createMemoryHistory } from 'history';
import StorageService from 'utils/StorageService';
import Emitter from 'utils/emitter';

import ClientHeader from '../ClientHeader';
import configureStore from '../../../configureStore';

describe('<ClientHeader />', () => {
  const history = createMemoryHistory();
  const store = configureStore({}, history);

  it('should render a div', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <ConnectedRouter history={history}>
            <ClientHeader />
          </ConnectedRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Test emitter functions', () => {
  test('Testing if toggle with proxyLoginClient = true', () => {
    const component = shallow(<ClientHeader />);
    const token = 'token';
    StorageService.set('jwtToken', token);
    StorageService.set('signupStep', 3);
    StorageService.set('proxyType', 'client');
    StorageService.set('userType', 2);
    Emitter.emit('proxyLoginClient', true);
    expect(component.state().proxyType).toEqual('client');
  });

  test('Testing if toggle with proxyLoginClient = false', () => {
    const component = shallow(<ClientHeader />);
    StorageService.set('proxyType', 'client');
    Emitter.emit('proxyLoginClient', false);
    expect(component.state().proxyType).toEqual('client');
  });

  test('Testing if toggle with proxyBackToAdmin = true', () => {
    const component = shallow(<ClientHeader />);
    StorageService.set('userType', 4);
    Emitter.emit('proxyBackToAdmin', true);
    expect(component.state().userType).toEqual(4);
  });

  test('Testing if toggle with proxyBackToAdmin = false', () => {
    const component = shallow(<ClientHeader />);
    StorageService.set('userType', 2);
    Emitter.emit('proxyBackToAdmin', false);
    expect(component.state().userType).toEqual(4);
  });
});

describe('Test functions', () => {
  const getWrapper = () => shallow(<ClientHeader />);
  const getInstance = () => getWrapper().instance();

  test('Testing if dropdownToggle', () => {
    const dropdownToggle = jest.spyOn(getInstance(), 'dropdownToggle');
    dropdownToggle();
    expect(dropdownToggle).toHaveBeenCalledTimes(1);
  });

  test('Testing if menuToggle', () => {
    const menuToggle = jest.spyOn(getInstance(), 'menuToggle');
    menuToggle();
    expect(menuToggle).toHaveBeenCalledTimes(1);
  });
});
