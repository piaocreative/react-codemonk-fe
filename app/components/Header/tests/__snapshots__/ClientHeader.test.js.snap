// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientHeader /> should render a div 1`] = `
<div
  class="header-style__HeaderNav-sc-196epht-4 fqsaJt"
>
  <nav
    class="header-style__FixedNavbar-sc-196epht-5 iUzMck client-header navbar"
  >
    <div
      class="header-style__NavbarContainer-sc-196epht-11 bieUll container-fluid"
    >
      <div
        class="d-flex align-items-center navbar-expand-md"
      >
        <a
          class="brand-style__NavBrand-kuuhq1-0 jDEDXe"
          href="/client"
          title="CodeMonk"
        >
          <img
            alt="CodeMonk"
            class="d-none d-md-block"
            src="IMAGE_MOCK"
          />
          <img
            alt="CodeMonk"
            class="d-block d-md-none"
            src="IMAGE_MOCK"
          />
        </a>
      </div>
    </div>
  </nav>
</div>
`;
